Coolix improvements

ribeirodanielf opened the pull request:
Hi David,

After your comments, seems I've found the less intrusive way to improve and fix some issues with the code. I'm testing it on my real setup and so far is working fine.
My apologies but I don't know yet how to execute the automated tests.
Could you please check the code?
Regards,

Daniel Ribeiro

Commit:
fix some issues and add flags for transient states

Commit:
while sending Fan command temperature goes to 0xf, force it to show the max temperature instead

Commit:
adjustment for automated tests

Commit:
missing some initialization and removing Raw test for transiente state

crankyoldgit commented:
> My apologies but I don't know yet how to execute the automated tests.

See https://github.com/crankyoldgit/IRremoteESP8266/wiki/Library-Maintainers-Guide#unit-tests for how to run them yourself locally.

> Could you please check the code?

Will do.

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -100,21 +101,43 @@ IRCoolixAC::IRCoolixAC(const uint16_t pin, const bool inverted,
   _irsend.sendCOOLIX(remote_state, kCoolixBits, repeat);
+  // make sure to remove special state from remote_state
+  // after command has being transmitted.
+  recoverSavedState();
```
Nice catch!

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -345,6 +404,7 @@ void IRCoolixAC::setFan(const uint8_t speed, const bool modecheck) {
       break;
     default:  // Unknown speed requested.
       newspeed = kCoolixFanAuto;
+      break;
```
FYI. This line isn't needed. You don't need a `break;` when it's the last line/statement of a `switch` statement.

ribeirodanielf replied:
Thought this is a good code practice to make the code clear.

crankyoldgit commented on test/ir_Coolix_test.cpp:
```diff
@@ -352,15 +352,17 @@ TEST(TestDecodeCoolix, FailToDecodeNonCoolixExample) {
   ircoolix.setRaw(kCoolixDefaultState);
   EXPECT_EQ(kCoolixDefaultState, ircoolix.getRaw());
-}
+}*/
```
Yes it is. `getRaw()` should always return what was just `setRaw()`ed. This test should pass.

`kCoolixOff` and other special cases may be different now. So change `kCoolixOff` to some other "normal" state.
This is to ensure `setRaw` & `getRaw` actually do what they advertise. :)

crankyoldgit reviewed (changes requested):
Much easier to read this time! Thanks!

Commit:
Off is a transient state, it requires Send to return back to regular state

Commit:
fixing code review comments

Commit:
more adjustments based on the test results

Commit:
must be on to run commands

Commit:
changed the common send message

Commit:
additional point

ribeirodanielf commented:
> > My apologies but I don't know yet how to execute the automated tests.
> 
> See https://github.com/crankyoldgit/IRremoteESP8266/wiki/Library-Maintainers-Guide#unit-tests for how to run them yourself locally.
> 
> > Could you please check the code?
> 
> Will do.

Hi David, I've changed most of the points you raised and fixed the tests but seems something else is wrong and I cannot figure out.
Travis is giving an error but I can see no red lines for 2672.2 build job.

Could you please check and advise.
Thanks

crankyoldgit commented:
There was a bunch of linter issues when I looked last time. Did you address those? Search for "exited with" in the Travis log.

crankyoldgit commented:
Now I'm near a keyboard again, here are the current issues causing the Travis checks to fail:

https://travis-ci.org/crankyoldgit/IRremoteESP8266/jobs/593345709#L441-L450
```
src/ir_Coolix.h:78:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:79:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:80:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:81:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:82:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:83:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:84:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:85:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:86:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Coolix.h:141:  Missing spaces around =  [whitespace/operators] [4]
```
&
https://travis-ci.org/crankyoldgit/IRremoteESP8266/jobs/593345709#L486-L487
```
src/ir_Coolix.cpp:131:  Missing space before {  [whitespace/braces] [5]
src/ir_Coolix.cpp:133:  Missing space before {  [whitespace/braces] [5]
```

They are just minor linter issues.

Commit:
minor adjustment to comply with linter

ribeirodanielf commented:
> Now I'm near a keyboard again, here are the current issues causing the Travis checks to fail:
> 
> https://travis-ci.org/crankyoldgit/IRremoteESP8266/jobs/593345709#L441-L450
> 
> ```
> src/ir_Coolix.h:78:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:79:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:80:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:81:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:82:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:83:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:84:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:85:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:86:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/ir_Coolix.h:141:  Missing spaces around =  [whitespace/operators] [4]
> ```
> 
> &
> https://travis-ci.org/crankyoldgit/IRremoteESP8266/jobs/593345709#L486-L487
> 
> ```
> src/ir_Coolix.cpp:131:  Missing space before {  [whitespace/braces] [5]
> src/ir_Coolix.cpp:133:  Missing space before {  [whitespace/braces] [5]
> ```
> 
> They are just minor linter issues.

My bad! I was so focused to solve the Test issues that I forgot to check linter :(
Just submitted the minor fixes.
Thanks

Commit:
make sure no more commands will be send after turn off coolix ac

crankyoldgit commented on src/IRac.cpp:
```diff
@@ -1905,6 +1910,7 @@ namespace IRAcUtils {
 #if DECODE_COOLIX
       case decode_type_t::COOLIX: {
         IRCoolixAC ac(0);
+        ac.on();
```
Why is this statement required? `setRaw()` should (in theory) generate the correct internal state for `setPower()` from `result->value`.
As I understand it (I don't have access to a "Coolix" compatible A/C system), every IR message that is NOT an "off" command is considered an "on" message. Is that a correct assumption?

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -132,6 +155,36 @@ bool IRCoolixAC::isSpecialState(void) {
 }
 
+// Special state means commands that are not
+// afecting Temperature/Mode/Fan
```
typo: "affecting"

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -170,11 +218,10 @@ uint8_t IRCoolixAC::getTemp() {
   for (i = 0; i < kCoolixTempRange; i++)
     if (kCoolixTempMap[i] == code) return kCoolixTempMin + i;
-  return kCoolixUnknown;  // Not a temp we expected.
+  return kCoolixTempMax;  // Not a temp we expected.
```
Might this hide when we are getting a value we don't (yet) understand, and perhaps should?

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -278,43 +344,41 @@ void IRCoolixAC::setMode(const uint8_t mode) {
     case kCoolixAuto:
     case kCoolixDry:
-      if (this->getFan() == kCoolixFanAuto)
-        //  No kCoolixFanAuto in Dry/Auto mode.
```
Is this comment/assumption incorrect? IIRC when I added this, the person said there was no "auto" fan speed when in dry/auto operation modes.
Can you confirm that isn't the case? If you can have fan set to auto in dry/auto operation mode, then this is obviously fine as is.

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -278,43 +344,41 @@ void IRCoolixAC::setMode(const uint8_t mode) {
-      if (this->getFan() == kCoolixFanAuto)
-        //  No kCoolixFanAuto in Dry/Auto mode.
-        this->setFan(kCoolixFanAuto0, false);
+        setFan(kCoolixFanAuto0, false);
```
As you've removed the `if` statement, you should remove one level of indent.

crankyoldgit replied:
fixed

crankyoldgit commented on src/ir_Coolix.h:
```diff
@@ -134,13 +138,26 @@ class IRCoolixAC {
 #else
   IRsendTest _irsend;
 #endif
+  enum State {sOff = 0, sOn = 1};
```
Why use an `enum` for this instead of just a `bool`?

crankyoldgit replied:
Changed.

crankyoldgit reviewed (approved):
Code wise, I'm fine with this. Just a few questions to clarify/confirm what you are proposing is correct.
Once confirmed, I'm happy to merge etc.

crankyoldgit commented:
@ribeirodanielf Could you please answer some of my code review questions?

crankyoldgit committed:
Merge branch 'master' into coolix_imp

crankyoldgit committed:
Cleanup after merge.

crankyoldgit committed:
Ammend code to address review comments.

crankyoldgit commented:
FYI, the changes mention above have been included in the v2.7.0 release of the library.

CRCinAU commented:
Just to be annoying, this seems to have broken the Off function at a bare minimum.

I'm trying to debug to see what the story is - but I currently can't turn off my coolix based unit :)

CRCinAU commented:
Ah - if I understand this right, I can no longer use ac.setpower(false) - but I have to use ac.off()

Is there any updated examples of the Coolix functionality now that I can check the rest of my code against?

crankyoldgit commented:
> Ah - if I understand this right, I can no longer use ac.setpower(false) - but I have to use ac.off()

In `ir_Coolix.cpp`:
```
void IRCoolixAC::off(void) { this->setPower(false); }
```

So all `ac.off()` is doing for Coolix is `setPower(false)` i.e. the should be interchangeable and identical.

> Is there any updated examples of the Coolix functionality now that I can check the rest of my code against?

Not sure what you mean, but I think we can address this in your new issue you've logged.
